### PR TITLE
CBMC additional profiling info: Pointer dereference, points-to set metric

### PR DESCRIPTION
**Pull request by natasha-jeppu**
Parse option to track size and contents of points-to sets during pointer dereferencing.
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/pointer-analysis/value_set_dereference.cpp
Comment (peterschrammel) [inline src/pointer-analysis/value_set_dereference.cpp]: 🚫 This is not allowed. Pass a message handler to the class instead, as done everywhere else.

**Review by martin-cs — COMMENTED** [review]
1. There are already a number of ways of showing the points-to sets in general.
2. If it must specifically be done during symbolic execution, why not just add it as a call to the right level of logging.  I don't see why we need extra options.

FYI : there has been a long-standing goal of minimising the number of options that CPROVER tools have or at least preventing the proliferation of them.  We haven't done so well of late but it is still a worthwhile goal to pursue.

**Comment by natasha-jeppu** [comment]
> 1. There are already a number of ways of showing the points-to sets in general.

Are you referring to the --show-value-sets option? As I understand, the CBMC call just looks at the points-to set information which is recorded during symex and this is what we are interested in.

**Comment by danielsn** [comment]
> 1. There are already a number of ways of showing the points-to sets in general.
> 2. If it must specifically be done during symbolic execution, why not just add it as a call to the right level of logging.  I don't see why we need extra options.
> 
> FYI : there has been a long-standing goal of minimising the number of options that CPROVER tools have or at least preventing the proliferation of them. We haven't done so well of late but it is still a worthwhile goal to pursue.

I wonder if it would make sense to have a mechanism for defining what statistics to collect / what output to generate, that could go under a single commandline flag.
Ideas are: 
1. A json config file
2. A `--show-metric-in-log metric` flag that takes a metric, or list of metrics.

File: src/pointer-analysis/value_set_dereference.cpp
Comment (natasha-jeppu) [reply src/pointer-analysis/value_set_dereference.cpp]: Fixed this.

**Comment by peterschrammel** [comment]
If it's just about showing _metrics_ that are cheap to collect (e.g. points to set size) then I would just always output them - no flag needed. Such metrics are indeed very useful.
If you are also going to show the points to sets themselves then I'd prefer to put them behind a flag, as this produces _massive_ output... and I'm not sure what it is good for except for very special debug situations.

**Comment by natasha-jeppu** [comment]
> If it's just about showing _metrics_ that are cheap to collect (e.g. points to set size) then I would just always output them - no flag needed. Such metrics are indeed very useful.
> If you are also going to show the points to sets themselves then I'd prefer to put them behind a flag, as this produces _massive_ output... and I'm not sure what it is good for except for very special debug situations.

For now both are behind a flag. We prefer to keep this behind a flag because even with just the points-to set size, we can get a really big output (multiple pointer dereferences and number of pointers may be large). But if you prefer to have it as part of the standard output we could make the change.

File: src/pointer-analysis/value_set_dereference.cpp
Comment (peterschrammel) [inline src/pointer-analysis/value_set_dereference.cpp:14]: ❗Please don't remove the ifdef here. `iostream` must not be included outside `DEBUG` here.

You can use `iosfwd` outside `DEBUG` - that's what you need here, as far as I can see.

File: src/pointer-analysis/goto_program_dereference.h
Comment (peterschrammel) [inline src/pointer-analysis/goto_program_dereference.h]: ❗ Passing a default `messaget` instance doesn't look right here. The instance should come from the caller.

Comment (natasha-jeppu) [reply src/pointer-analysis/goto_program_dereference.h]: Is this a valid construct? The points-to sets are displayed during symex only. `exprt dereference(const exprt &pointer, bool display_points_to_sets = false);` sets `display_points_to_set` to false by default so any calls to `value_set_dereferencet::dereference()` from the goto program conversion phase does not display the points-to sets. Similarly, is it okay to set the `messaget` here to a default message instance for `goto_program_dereferencet`?
```
goto_program_dereferencet(
    const namespacet &_ns,
    symbol_tablet &_new_symbol_table,
    const optionst &_options,
    value_setst &_value_sets,
    const messaget &_log = messaget())
    : options(_options),
      ns(_ns),
      value_sets(_value_sets),
      dereference(_ns, _new_symbol_table, *this, ID_nil, false, _log)
  {
  }
```

Passing a valid messaget object here`goto_program_dereferencet` will need a lot of additional refactoring across the goto program conversion code modules.

Comment (peterschrammel) [reply src/pointer-analysis/goto_program_dereference.h]: Ok, thanks for the explanation. Please put a code comment to clarify why a default messaget instance is passed in this place. The next person stumbling over that code should not have to redo your reasoning. Many thanks!

**Review by peterschrammel — APPROVED** [review]

**Comment by peterschrammel** [comment]
Approved assuming that the last two comments will be fixed before merging.

File: src/pointer-analysis/goto_program_dereference.h
Comment (natasha-jeppu) [reply src/pointer-analysis/goto_program_dereference.h]: Should I retain the current implementation i.e `dereference(_ns, _new_symbol_table, *this, ID_nil, false, messaget())` or replace it with the construct above? 
I will add in the comment.

Comment (peterschrammel) [reply src/pointer-analysis/goto_program_dereference.h]: The one above is probably slightly cleaner as it pushes the messaget further up. But, still, please put a comment in goto_program_dereferencet then.

Comment (natasha-jeppu) [reply src/pointer-analysis/goto_program_dereference.h]: Fixed. Comment added.

File: src/pointer-analysis/value_set_dereference.cpp
Comment (natasha-jeppu) [reply src/pointer-analysis/value_set_dereference.cpp:14]: Fixed.

**Pull request merged**
